Src/FluentAssertions/Primitives/ReferenceTypeAssertions.cs

```diff
@@ -416,6 +417,44 @@ public AndConstraint<TAssertions> NotBeAssignableTo(Type type, string because =
         return new AndConstraint<TAssertions>((TAssertions)this);
     }
 
+    /// <summary>
+    /// Asserts that the criteria provided by the element inspector is satisfied.
+    /// </summary>
+    /// <param name="expected">The element inspector which must be satisfied by the <typeparamref name="TSubject" />.</param>
+    /// <returns>An <see cref="AndConstraint{T}" /> which can be used to chain assertions.</returns>
+    /// <exception cref="ArgumentNullException"><paramref name="expected"/> is <see langword="null"/>.</exception>
+    public AndConstraint<TAssertions> Satisfy<T>(Action<T> expected)
+        where T : TSubject
+    {
+        Guard.ThrowIfArgumentIsNull(expected, nameof(expected), "Cannot verify an object against a <null> inspector.");
+
+        Execute.Assertion
+            .ForCondition(Subject is T)
+            .WithDefaultIdentifier(Identifier)
+            .FailWith("Expected {context:object} to be assignable to {0}{reason}, but {1} is not.", typeof(T), Subject.GetType());
+
+        string[] failuresFromInspector;
+
+        using (var assertionScope = new AssertionScope())
+        {
+            expected((T)Subject);
+            failuresFromInspector = assertionScope.Discard();
+        }
+
+        if (failuresFromInspector.Length > 0)
```

Comment:
🤔 Do we even need to do this? What if you just relied on the assertion scope to throw a combined message? What kind of information would you be missing?

Reply:
I don't know, I'll test it to see if it's necessary 👍

Reply:
And?

Reply:
Sorry, I haven't had time to do it yet (I'm working on this in my company's "personal time", which is only every second week).
I will look into it this evening.

Reply:
I tested the changes you suggested, and it looks like we will be losing the outer object context. There is also no indentation on the nested assertions, which I think is useful.

I don't know what is preferred though, however, this is how it's done in `Match`, and I believe the result should be the same or similar, no?

Using a single assertion scope:
```
Expected person.Name to be the same string, but they differ at index 1:
    ↓ (actual)
  "Buford Howard Tannen"
  "Biff Tannen"
    ↑ (expected).
Expected person.Age to be 48, but found 87 (difference of 39).
Expected address.City to be "Hill Valley, San Diego County, California" with a length of 41, but "Hill Valley" has a length of 11, differs near "y" (index 10).
```

Using `Discard`:
```
Expected complexDto to match inspector, but the inspector was not satisfied:
	Expected dto.Person to match inspector, but the inspector was not satisfied:
		Expected person.Name to be the same string, but they differ at index 1:
		    ↓ (actual)
		  "Buford Howard Tannen"
		  "Biff Tannen"
		    ↑ (expected).
		Expected person.Age to be 48, but found 87 (difference of 39).
	Expected dto.Address to match inspector, but the inspector was not satisfied:
		Expected address.City to be "Hill Valley, San Diego County, California" with a length of 41, but "Hill Valley" has a length of 11, differs near "y" (index 10).
```

Reply:
When I rely solely on the assertion scope, I get this

```
Expected dto.Person to be the same string, but they differ at index 1:
    ↓ (actual)
  "Buford Howard Tannen"
  "Biff Tannen"
    ↑ (expected).
Expected dto.Person to be 48, but found 87 (difference of 39).
Expected dto.Address to be "Hill Valley, San Diego County, California" with a length of 41, but "Hill Valley" has a length of 11, differs near "y" (index 10).
```

I personally feel like this is much more readable without all the noisy mentioning of the inspector (which is the wrong term anyway), don't you agree?

Reply:
Something is different in your case. It's not the Person or Address properties of the dto that failed, it's the person.Name and address.City. I think the assertion is presenting the wrong thing.
However, if it's at least showing the correct property in context of the parent, I believe either case is fine. Again, this is based on the implementation Match, so I'm not really trying to make any decisions, I'm just attempting to following an existing pattern 🙂

Reply:
Just so we talk about the same thing. Which test did you use to reproduce this scenario?

Reply:
The `Complex_typed_object_not_satisfying_inspector_throws` test.

Reply:
What's the conclusion on this one?

Reply:
I need to investigate this one a bit better, but haven't had any free time this week. Will try to take a look this weekend.

Reply:
I tried again and that made me realize that I don't like the parts about `Expected complexDto to match inspector, but the inspector was not satisfied:`. My suggestion is to change the implementation to 

```csharp
        using (new AssertionScope(CallerIdentifier.DetermineCallerIdentity()))
        {
            assertion((T)Subject);
        }

        return new AndConstraint<TAssertions>((TAssertions)this);
```

This will make your test result in 

```
Expected dto.Person to be the same string, but they differ at index 1:
    ↓ (actual)
  "Buford Howard Tannen"
  "Biff Tannen"
    ↑ (expected).
Expected dto.Person to be 48, but found 87 (difference of 39).
Expected dto.Address to be "Hill Valley, San Diego County, California" with a length of 41, but "Hill Valley" has a length of 11, differs near "y" (index 10).
```

I'll create a separate PR to make sure that nested calls to `AssertionScope(string)` will chain the context you provide so that you'll get something like 

```
Expected complexDto/dto.Person to be the same string, but they differ at index 1:
    ↓ (actual)
  "Buford Howard Tannen"
  "Biff Tannen"
    ↑ (expected).
Expected complexDto/dto.Person to be 48, but found 87 (difference of 39).
Expected complexDto/dto.Address to be "Hill Valley, San Diego County, California" with a length of 41, but "Hill Valley" has a length of 11, differs near "y" (index 10).
```

Reply:
Created #2607

Reply:
```
Expected complexDto/dto.Person to be the same string, but they differ at index 1:
    ↓ (actual)
  "Buford Howard Tannen"
  "Biff Tannen"
    ↑ (expected).
Expected complexDto/dto.Person to be 48, but found 87 (difference of 39).
```

I think what @siewers meant was, that this message presenting the wrong "objects" on which the assertion fails.
Both failing on `Person` but the first mentioning a `string` which doesn't match, and the latter suddenly mentioning (for the same property) an `int` which is failing.

Reply:
When you rebase this PR on `develop` (and thereby include #2607), you'll get this:

```
Expected complexDto to match inspector, but the inspector was not satisfied:
	Expected dto.Person to match inspector, but the inspector was not satisfied:
		Expected person.Name to be the same string, but they differ at index 1:
		    ↓ (actual)
		  "Buford Howard Tannen"
		  "Biff Tannen"
		    ↑ (expected).
		Expected person.Age to be 48, but found 87 (difference of 39).
	Expected dto.Address to match inspector, but the inspector was not satisfied:
		Expected address.City to be "Hill Valley, San Diego County, California" with a length of 41, but "Hill Valley" has a length of 11, differs near "y" (index 10).
```

It is still a bit noisy because of the superfluous "Expected complexDto to match inspector, but the inspector was not satisfied", but I'm pretty happy with the result.

Reply:
I've rebased now, but am I supposed to then only the following in `Satisfy`?
```cs
using (new AssertionScope(CallerIdentifier.DetermineCallerIdentity()))
{
    assertion((T)Subject);
}
```
If I do that, I'm getting this assertion message, which isn't the same as you wrote (and doesn't really make sense) :)
```
Expected complexDto/dto.Person to be the same string, but they differ at index 1:
    ↓ (actual)
  "Buford Howard Tannen"
  "Biff Tannen"
    ↑ (expected).
Expected complexDto/dto.Person to be 48, but found 87 (difference of 39).
Expected complexDto/dto.Address to be "Hill Valley, San Diego County, California" with a length of 41, but "Hill Valley" has a length of 11, differs near "y" (index 10).
```

Reply:
The problem with the message is that it doesn't mention the failed nested property names:
- `complexDto.Person.Name`
- `complexDto.Person.Age`
- `complexDto.Address.City`
If those exact property names aren't mentioned in the message, it's not helpful.

As it looks with the new changes, both the `Name` and `Age` properties gets merged into a single nonsensical error on the same `Person` property.
It's wrong to say that `Person` was expected to be the same string _and_ it was expected to be a certain number :)
Also, `Address` is another "complex" object, but the error message doesn't present the exact property that failed the assertion.

I hope that makes sense :)

Reply:
I need to look into that (again). Later today.

Reply:
Crap, you're right. The call to `new AssertionScope(CallerIdentifier.DetermineCallerIdentity())` doesn't actually help at all. Sorry about that. I thought I could fix that by changing some of the behavior `AssertionScope` uses. But it turns out, it doesn't. So maybe we can rewrite `Expected dto.Person to match inspector, but the inspector was not satisfied:` to be less verbose.

Reply:
Please also note that `CallerIdentifier.DetermineCallerIdentity()` is very expensive to call and is only needed when the test fails.
See #1660 and #1657

Reply:
> Please also note that `CallerIdentifier.DetermineCallerIdentity()` is very expensive to call and is only needed when the test fails.

True, but it's no longer relevant. My suggestion doesn't work, so the original implementation is still the best.